### ISSUE TITLE
Support for Initial State Estimation in Neural MPC with RNN Models

### ISSUE THREAD

**Issue by L-3K** [FilippoAiraldi/casadi-nlp#3]
I am working on creating a Neural Model Predictive Controller (MPC) using `csnlp.wrappers.mpc` within the `csnlp` library. My trained PyTorch LSTM model is converted into a CasADi function using `l4casadi`, which I use as the prediction model in the MPC via the `set_nonlinear_dynamics` method.

One critical challenge is related to the initial state of the RNN (e.g., hidden and cell states of the LSTM). Typically, the initial state of an RNN is set to zero, but this can lead to inaccurate transient predictions, especially during the first few steps of the prediction horizon. Capturing these early steps accurately is crucial for applications like MPC. 

Estimating the initial state is important for improving prediction accuracy and I would like to extend this idea.

### Request/Question:
- Does the `csnlp` library currently support the estimation of the initial hidden/cell states in RNN models (such as LSTM) for MPC, particularly using a past window of `M` previous outputs and control signals to estimate the initial state?
- If not, is there a recommended approach or workaround within the `csnlp` framework to implement this feature? Specifically, how can I open the prediction loop for the first `M` steps and incorporate previous measurements and control signals to estimate the initial hidden and cell states before running the MPC?

This feature is crucial for ensuring accurate initial predictions in Neural MPC applications. Any guidance or suggestions would be greatly appreciated.

Thank you for your help!

**Comment by FilippoAiraldi** [comment]
Hey @LOPES3000,

To answer your questions

* `csnlp` does not provide state estimation capabilities. This is true for your case, but also for more classical control frameworks, e.g., nominal MPC with Kalman filter: the package simply does not provide filters/estimators. 
* That been said, the package could very well support estimation. You just need to craft an estimator that, given some data, yields an estimated state of the system, and feed this estimate to the `mpc` (as something like `mpc.solve(pars={"x_0": your_estimate_of_x_0})`, assuming your state is called `"x"`).

I feel like you are basing your questions on papers like [this one](https://arxiv.org/pdf/2201.08660) or [this other one](https://www.sciencedirect.com/science/article/pii/S0947358024001080). I am not an expert on this topic, but here is my two cents.

When the state is fully observed (i.e., there is no need to estimate the state), a simulation loop would look like this:

```python
mpc = ...  # assume we have used csnlp to create an mpc controller
env.reset()  # reset the environment/system (OpenAI Gym-like)
for t in range(t_final):
    x_real = env.x  # get the real environment's state
    sol = mpc.solve(pars={"x_0": x_real}  # solve the mpc optimization
    u_opt = sol.vals["u"][:, 0]  # take the first optimal action
    env.step(u_opt)  # apply it in a receding horizon fashion
```

This is pretty standard, of course. If the system's state cannot be measured, we need some sort of estimator

```python
mpc = ...
estimator = ...  # e.g., Kalman filter
env.reset()
for t in range(t_final):
    estimator.update(  # update the estimator with latest data
        env.input,
        env.output,
    )  
    x_estimated = estimator.get_estimate()  # compute the estimate
    sol = mpc.solve(pars={"x_0": x_estimated}  # the rest is as usual
    u_opt = sol.vals["u"][:, 0]
    env.step(u_opt)
```

With your RNN model, it would be a bit more complex I guess.

```python
rnn = ...  # instantiate the RNN estimator
rnn.train()  # train it on some offline dataset

# before simualting, warm-start the rnn initial state
env.reset() 
real_input_output_data = env.collect_io_data()
rnn.warm_start(real_input_output_data)

# now we can go for the online simulation
mpc = ...
for t in range(t_final):
    x_estimated = rnn(  # pass the input-output through the RNN to get an estimate
        env.input,
        env.output,
    )  
    x_estimated = estimator.get_estimate()  # the rest is as usual
    sol = mpc.solve(pars={"x_0": x_estimated}  
    u_opt = sol.vals["u"][:, 0]
    env.step(u_opt)
```

The warm-starting procedure is as you described, i.e., it uses a window of input-output data to estimate the hidden state of the RNN. However, it is not clear to me if this procedure must be done only at the beginning, or it can also be done at some frequency, e.g., every `N >> M` time steps the RNN initial state is re-calibrated.

By the way, if you are interested in building the RNN directly in CasADi (instead of using `l4casadi`'s method based on callbacks), I wrote a [package](https://github.com/FilippoAiraldi/casadi-neural-nets) to do so. It does not contain LSTMs yet, but all the building blocks should be there. Also, if you want to apply RL to fine-tune the RNN as [this paper](https://www.sciencedirect.com/science/article/pii/S0947358024001080) did, you can use another package of mine, `mpcrl`.

Bests!

**Comment by L-3K** [comment]
Hi @FilippoAiraldi,

Thank you for your reply!

You're correct; I'm trying to reproduce the Neural MPC controller from [this paper](https://www.sciencedirect.com/science/article/pii/S0947358024001080) using the `csnlp` and `mpcrl` libraries. However, since I'm struggling to solve the MPC for one step, I'm currently focusing on running the MPC without the RL algorithm.

The state of the system (environment) is fully observed. However, because my system dynamics are represented by an RNN network, I need to estimate the RNN's initial state. To achieve this, I use the model itself as a state estimator by sequentially feeding it a window of past measurements and outputs (which I call the "context"). This approach is better explained in [this paper](https://arxiv.org/pdf/2201.08660).

![RNN Initial State Estimation](https://github.com/user-attachments/assets/019f46b0-d678-456c-a528-08f2b70c1662)

After reading your reply, I came up with the following (partial) solution:

```python
# Assume I've used csnlp to create the MPC controller
mpc = ...

model = OpenLSTM(self.n_context, self.n_inputs, self.batch_size, self.sequence_length, is_estimator=True)
model.load_state_dict(torch.load(os.path.join("models", model_filename), map_location="cpu"))
model_l4c = l4c.L4CasADi(model_wrapped, device='cpu', batched=False)

input_sym = cs.MX.sym('input', (self.n_inputs + self.n_outputs), (self.horizon + self.n_context))
y_sym = model_l4c(input_sym)
f = cs.Function('y', [input_sym], [y_sym])

# Set dynamics
self.set_dynamics(f)
```

```python
# Warm-start the RNN initial state by feeding it with n_context collected states
env.reset()
for t in range(mpc.n_context):
    new_state, _, _, _, _ = env.step(u_context)
    x_context.append(new_state[0])

x_context_np = np.asarray(x_context).reshape(1, -1)

# Initiate the control loop
for t in range(t_final):
    # Get the last n_context states
    sol = mpc.solve(pars={"x_0": x_context_np[:, -(mpc.n_context):]})
    u_opt = sol.vals["u"][:, mpc.n_context]
    new_state, _, _, _, _ = env.step(u_opt)
    x_context.append(new_state[0])
    x_context_np = np.asarray(x_context).reshape(1, -1)
```

Internally, the CasADi function works like this:

- `input_sym` is a stacked array of states (`x`) and control signals (`u`).
- The length of `input_sym` is equal to the prediction horizon `N` plus `n_context`.
- On every internal call for the MPC dynamics, the function takes the first `n_context` column elements to estimate the RNN initial state.

To implement this, I needed to make some adaptations to `mpc.py` so that I can:

- Use "neural dynamics."
- Set an `n_context` for internal use.
- Ensure that the first actions retrieve the position of `self._n_context` of the controller.
- Adjust shapes:
  - `x` now has a shape of `(size, self._prediction_horizon + self._n_context)`.
  - `x0` now has a shape of `(size, self._n_context)`.
  - `u` now has a shape of `(size, self._prediction_horizon + self._n_context)`.

**PS:** I am using the following versions:

- `csnlp` version 1.6.0
- `mpcrl` version 1.2.1
- `l4casadi` version 2.0.0

You can find the modified file [here](https://github.com/LOPES3000/neural_mpc/blob/main/mpc.py) and my code [here](https://github.com/LOPES3000/neural_mpc/blob/main/NEURAL_MPC.ipynb).

The problem I'm facing now is that in the first steps, the solver converges saying "Solved to acceptable level," but after a few steps, it says: **Restoration failed**.

Do you have any suggestions for solving this issue? 

Thank you again!

**Comment by FilippoAiraldi** [comment]
Hi @LOPES3000,

it is not clear to me why you had to include the context states as variables in the MPC optimization problem, but again I am not an expert on this topic. 

To my understanding, an RNN model is a dynamical predictor of the form $x_{k+1} = f_{RNN}(x_k, u_k, h_k)$, where $x_k$ and $u_k$ are as usual states and inputs, and $h_k$ is the hidden state of the recurrent net. I thought the paper we both references talked about using this context window to properly initialize $h_0$ **before** the first MPC solver iteration at $k=0$. So I was surprised to see that you basically included this window in the optimization itself as a variable. 

Anyway, `ipopt` can be a bit difficult to tune, especially for highly nonlinear problems. As far as I know, the restoration phase is linked to the search for a feasible solution. Do you perhaps have bounds on the states? Maybe the solver is struggling to find a solution that is feasible, let alone optimal. You could rewrite these bounds as soft constraints via slack variables. Other than that, try to give a look at `ipopt` documentation and see if there's any option you can tweak to make it more robust. Lately, a new faster interior point solver (`fatrop`) is being shipped with casadi >= 3.6.7; you could also try that. I know for a fact it does not implement a restoration phase, so it might either fail completely or output an infeasible solution by default.
